flake/dev/public-and-dev-version-consistency.nix

```diff
@@ -2,40 +2,36 @@
 {
   perSystem =
     { lib, pkgs, ... }:
-    {
-      checks = lib.mkMerge (
-        map
-          (
-            input:
-            let
-              name = "${input}-and-dev-${input}-version-consistency";
-            in
-            {
-              ${name} =
-                let
-                  dev = inputs."dev-${input}".narHash;
-                  public = inputs.${input}.narHash;
-                in
-                pkgs.runCommand name { } ''
-                  if [ "${public}" != "${dev}" ]; then
-                    printf \
-                      'inconsistent ${input} (%s) and dev-${input} (%s) versions\n' \
-                      "${public}" \
-                      "${dev}" \
-                      >&2
+    let
+      checks = lib.genAttrs' [ "flake-parts" "nixpkgs" "systems" ] (
+        input:
+        let
+          name = "${input}-and-dev-${input}-version-consistency";
+        in
+        lib.nameValuePair name (
+          let
+            dev = inputs."dev-${input}".narHash;
+            public = inputs.${input}.narHash;
+          in
+          pkgs.runCommand name { } ''
+            if [ "${public}" != "${dev}" ]; then
+              printf \
+                'inconsistent ${input} (%s) and dev-${input} (%s) versions\n' \
+                "${public}" \
+                "${dev}" \
+                >&2
 
-                    exit 1
-                  fi
+              exit 1
+            fi
 
-                  mkdir "$out"
-                '';
-            }
-          )
-          [
-            "flake-parts"
-            "nixpkgs"
-            "systems"
-          ]
+            mkdir "$out"
+          ''
+        )
       );
+    in
+    {
+      inherit checks;
+
+      ci.buildbot.public-and-dev-version-consistency = pkgs.linkFarm "public-and-dev-version-consistency" checks;
```

Comment:
Another optimization, this only needs to be built on one platform. Since all inputs to these checks are platform agnostic.

I'd do something like `mkIf (system == "x86_64-linux")` or `"aarm64-linux"` depending on if you want to optimize for speed or power-efficiency.

Might need to do the `mkIf` on the `ci.buildbot` instead of the `public-and-dev-version-consistency`, because of `lazyAttrsOf`'s emptyValue. Not sure.

EDIT: or alternatively move this out of `perSystem` and define `flake.ci.buildbot.x86_64-linux.foobar`, no `mkIf` required... But then it's harder to get a `pkgs` instance I guess.

---

You could make a similar argument for `treefmt`, but the counter argument is that in that case you want to verify the test itself still works on all platforms.

Reply:
Assuming Buildbot spins one worker per `ci.buildbot` derivation, the worker number converges to the number of testbed groups, meaning this type of optimization might be negligible and make our implementation far more complex for practically no benefit. Similarly, CPU time would converge to the testbed evaluations.

Reply:
I agree we aren't talking about _big_ numbers, but there is no reason to spin up 4 workers to verify matching inputs when it can be done with 1 worker.

`mkIf (system == "aarch64-linux")` has negligible complexity cost, so I don't think that's a significant factor.

Reply:
> I agree we aren't talking about _big_ numbers, but there is no reason to spin up 4 workers to verify matching inputs when it can be done with 1 worker.
>
> `mkIf (system == "aarch64-linux")` has negligible complexity cost, so I don't think that's a significant factor.

Agreed. However, this allows confusing edge cases in the future when the invariant that all systems test the same thing no longer holds for some reason.

Still, I would be fine with this. But there should be a comment explaining that we assume all systems are functionality equivalent and that this reduces the total number of tests.